src/kv/memstore.rs

```diff
@@ -108,76 +113,147 @@ impl Write for WriteTransaction<'_> {
+                Some(v) => self.guard.insert(item.0.clone(), v.clone()),
+                None => self.guard.remove(item.0),
             };
         }
         Ok(())
     }
 
     async fn rollback(self: Box<Self>) -> Result<()> {
+        // TODO(phritz): determine what happens to async calls pending on
+        // a lock that has been dropped (eg, the pending mutex).
         Ok(())
     }
 }
 
+// TODO these tests should run against a Store trait object so we can test all types of
+// stores, not just the memstore.
 #[cfg(not(target_arch = "wasm32"))]
 #[cfg(test)]
 mod tests {
     use super::*;
     use crate::kv::StoreError;
 
     #[async_std::test]
-    async fn basics() -> std::result::Result<(), StoreError> {
+    async fn test_store_trait() -> std::result::Result<(), StoreError> {
         let mut ms = MemStore::new();
+
+        // Test put/has/get, which use read() and write() for one-shot txs.
         assert_eq!(false, ms.has("foo").await?);
         assert_eq!(None, ms.get("foo").await?);
+
         ms.put("foo", "bar".as_bytes()).await?;
         assert_eq!(true, ms.has("foo").await?);
         assert_eq!(Some("bar".as_bytes().to_vec()), ms.get("foo").await?);
-        assert_eq!(false, ms.has("bar").await?);
-        assert_eq!(None, ms.get("bar").await?);
 
-        let rt = ms.read().await?;
-        assert_eq!(false, rt.has("bar").await?);
-        assert_eq!(None, rt.get("bar").await?);
+        ms.put("foo", "baz".as_bytes()).await?;
+        assert_eq!(true, ms.has("foo").await?);
+        assert_eq!(Some("baz".as_bytes().to_vec()), ms.get("foo").await?);
 
-        let wt = ms.write().await?;
-        assert_eq!(false, wt.has("bar").await?);
-        wt.put("bar", b"baz").await?;
-        assert_eq!(Some(b"baz".to_vec()), wt.get("bar").await?);
-        assert_eq!(None, rt.get("bar").await?); // Test isolation.
-        wt.commit().await?;
+        assert_eq!(false, ms.has("baz").await?);
+        assert_eq!(None, ms.get("baz").await?);
+        ms.put("baz", "bat".as_bytes()).await?;
+        assert_eq!(true, ms.has("baz").await?);
+        assert_eq!(Some("bat".as_bytes().to_vec()), ms.get("baz").await?);
+
+        Ok(())
+    }
+
+    #[async_std::test]
+    async fn test_read_transaction() -> std::result::Result<(), StoreError> {
+        let mut ms = MemStore::new();
+        ms.put("k1", "v1".as_bytes()).await?;
 
         let rt = ms.read().await?;
-        assert_eq!(true, rt.has("bar").await?);
-        assert_eq!(Some(b"baz".to_vec()), rt.get("bar").await?);
+        assert_eq!(true, rt.has("k1").await?);
+        assert_eq!(Some(b"v1".to_vec()), rt.get("k1").await?);
 
         Ok(())
     }
 
     #[async_std::test]
-    async fn delete() -> std::result::Result<(), StoreError> {
+    async fn test_write_transaction() -> std::result::Result<(), StoreError> {
         let mut ms = MemStore::new();
-        ms.put("bar", "foo".as_bytes()).await?;
+        ms.put("k1", "v1".as_bytes()).await?;
```

Comment:
[minifix] `b"v1"` instead of `"v1".to_bytes()` -> clippy will tell you this too.